File: app/src/main/java/com/waitty/waiter/activity/LoginActivity.java
package com.waitty.waiter.activity;

import android.app.NotificationManager;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.waitty.waiter.R;
import com.waitty.waiter.appinterface.getResponseData;
import com.waitty.waiter.constant.constant;
import com.waitty.waiter.databinding.ActivityLoginBinding;
import com.waitty.waiter.fcm.RegisteFCMId;
import com.waitty.waiter.retrofit.API;
import com.waitty.waiter.retrofit.APICall;
import com.waitty.waiter.retrofit.ApiClient;
import com.waitty.waiter.retrofit.ApiInterface;
import com.waitty.waiter.utility.MyLoading;
import com.waitty.waiter.utility.Utility;
import org.json.JSONObject;
import retrofit2.Call;

public class LoginActivity extends BaseActivity implements getResponseData {

    private Context mContext;
    private ActivityLoginBinding activityLoginBinding;
    private ClickHandler mclickHandler;
    private MyLoading loader;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left);
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE);
        activityLoginBinding= DataBindingUtil.setContentView(this, R.layout.activity_login);
        mContext = this;
        init();
    }

    // Variable initialization
    private void init() {
        loader = new MyLoading(mContext);
        new RegisteFCMId(mContext).execute();
        mclickHandler = new ClickHandler();
        activityLoginBinding.setClickEvent(mclickHandler);

        activityLoginBinding.edtWaiterID.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void afterTextChanged(Editable editable) {
                checkConditions();
            }
        });

        activityLoginBinding.edtPassword.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
            }
            @Override
            public void afterTextChanged(Editable editable) {
                checkConditions();
            }
        });
    }

    // Click event handler
    public class ClickHandler {

        public ClickHandler( ) {
        }

        public void loginClick(View view) {
            NotificationManager nMgr = (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
            nMgr.cancelAll();
            if(Utility.isNetworkAvailable(mContext)) {
                if (Utility.getSharedPreferencesString(mContext, constant.USER_FCMTOKENID).trim().length() > 0 && Utility.getSharedPreferencesString(mContext, constant.USER_DEVICEID).trim().length() > 0)
                    loginAPI();
                else
                    new RegisteFCMId(mContext).execute();
            }else
                Utility.ShowSnackbar(mContext, activityLoginBinding.Cordinator, getString(R.string.check_network));
        }

    }

    // Check form conditions
    private void checkConditions() {
        String waiterID=activityLoginBinding.edtWaiterID.getText().toString().trim();
        String password=activityLoginBinding.edtPassword.getText().toString().trim();

        boolean waiterIDConditions=false;
        if(waiterID.length()==0) {
            activityLoginBinding.edtWaiterID.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.name_icon, 0, 0, 0);
            activityLoginBinding.txtErrorWaiterID.setVisibility(View.GONE);
        }else if(waiterID.length()<5){
            activityLoginBinding.edtWaiterID.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.name_icon_red, 0, 0, 0);
            activityLoginBinding.edtWaiterID.setTextColor(getResources().getColor(R.color.colorValidation));
            activityLoginBinding.txtErrorWaiterID.setVisibility(View.VISIBLE);
        }else{
            activityLoginBinding.edtWaiterID.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.name_icon_black, 0, R.mipmap.done_icon, 0);
            activityLoginBinding.edtWaiterID.setTextColor(getResources().getColor(R.color.colorNextWelcome));
            activityLoginBinding.txtErrorWaiterID.setVisibility(View.GONE);
            waiterIDConditions=true;
        }

        boolean passwordConditions=false;
        if(password.length()==0) {
            activityLoginBinding.edtPassword.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.password_icon, 0, 0, 0);
            activityLoginBinding.txtErrorPassword.setVisibility(View.GONE);
        }else if(password.length()<5){
            activityLoginBinding.edtPassword.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.password_icon_red, 0, 0, 0);
            activityLoginBinding.edtPassword.setTextColor(getResources().getColor(R.color.colorValidation));
            activityLoginBinding.txtErrorPassword.setVisibility(View.VISIBLE);
        }else{
            activityLoginBinding.edtPassword.setCompoundDrawablesWithIntrinsicBounds(R.mipmap.password_icon_black, 0, R.mipmap.done_icon, 0);
            activityLoginBinding.edtPassword.setTextColor(getResources().getColor(R.color.colorNextWelcome));
            activityLoginBinding.txtErrorPassword.setVisibility(View.GONE);
            passwordConditions=true;
        }

        if(waiterIDConditions && passwordConditions){
            activityLoginBinding.btnDiseable.setVisibility(View.INVISIBLE);
            activityLoginBinding.btnEnable.setVisibility(View.VISIBLE);
        }else{
            activityLoginBinding.btnDiseable.setVisibility(View.VISIBLE);
            activityLoginBinding.btnEnable.setVisibility(View.INVISIBLE);
        }
    }

    // Login API
    private void loginAPI() {
        try{
            loader.show(getString(R.string.wait));
            JsonObject jsonObject=new JsonObject();
            jsonObject.addProperty(API.KEY, activityLoginBinding.edtWaiterID.getText().toString().trim());
            jsonObject.addProperty(API.PASSWORD, activityLoginBinding.edtPassword.getText().toString().trim());
            jsonObject.addProperty(API.DEVICE_TYPE, constant.DEVICE_TYPE);
            jsonObject.addProperty(constant.USER_DEVICEID,  Utility.getSharedPreferencesString(mContext,constant.USER_DEVICEID));
            jsonObject.addProperty(constant.USER_FCMTOKENID, Utility.getSharedPreferencesString(mContext,constant.USER_FCMTOKENID));

            ApiInterface apiInterface = ApiClient.getClient().create(ApiInterface.class);
            Call<JsonElement> call = apiInterface.login(jsonObject);
            new APICall(mContext).Server_Interaction(call,this ,API.LOGIN);

        }catch (Exception e){
            e.printStackTrace();
            loader.dismiss();
        }

    }

    @Override
    public void onSuccess(JSONObject OBJ, String msg, String typeAPI) {

        try {
            loader.dismiss();
            if(OBJ.length()>0 && OBJ!=null){

                switch (typeAPI) {
                    case API.LOGIN:
                        Utility.ShowToast(mContext,msg,0);
                        Utility.setSharedPreferencesString(mContext, constant.USER_INFORMATION, OBJ.getJSONObject(API.DATA).toString());
                        Utility.setSharedPreferencesBoolean(mContext, constant.NOTIFICATIONS_SHOW, OBJ.getJSONObject(API.DATA).getBoolean(API.NOTIFICATION_TOGGLE));
                        startActivity(new Intent(mContext, HomeActivity.class));
                        finish();
                        break;
                }
            }

        }catch (Exception e){e.printStackTrace();}

    }

    @Override
    public void onFailed(String msg, String typeAPI) {
        loader.dismiss();
        if(!msg.isEmpty()){
            switch (typeAPI) {
                case API.LOGIN:
                    Utility.ShowSnackbar(mContext, activityLoginBinding.Cordinator,msg);
                    break;
            }
        }
    }

}

File: app/src/main/java/com/waitty/waiter/activity/LoginActivity.kt
package com.waitty.waiter.activity

import android.content.Context
import android.content.Intent
import android.content.res.ColorStateList
import android.graphics.Color
import android.graphics.drawable.Drawable
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.view.View
import android.view.Window
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.content.res.AppCompatResources
import androidx.core.content.ContextCompat
import androidx.core.text.buildSpannedString
import androidx.core.text.color
import androidx.databinding.DataBindingUtil
import com.waitty.waiter.R
import androidx.lifecycle.Observer
import com.google.android.material.textfield.TextInputEditText
import com.google.android.material.textfield.TextInputLayout
import com.translabtechnologies.visitormanagementsystem.vmshost.database.SharedPreferenceManager
import com.waitty.kitchen.fragment.FragmentUtils
import com.waitty.waiter.constant.constant
import com.waitty.waiter.databinding.ActivityLoginBinding
import com.waitty.waiter.model.APIStatus
import com.waitty.waiter.model.apimodels.LoginResponse
import com.waitty.waiter.retrofit.API
import com.waitty.waiter.utility.Utility
import com.waitty.waiter.utility.WKClickListener
import com.waitty.waiter.viewmodel.ApiErrorViewModel
import com.waitty.waiter.viewmodel.ClickType
import com.waitty.waiter.viewmodel.LoginViewModel
import androidx.databinding.library.baseAdapters.BR

class LoginActivity : AppCompatActivity(), WKClickListener {
    private lateinit var activityLoginBinding: ActivityLoginBinding
    private lateinit var loginViewModel: LoginViewModel
    private lateinit var apiErrorViewModel: ApiErrorViewModel


    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        overridePendingTransition(R.anim.slide_in_right, R.anim.slide_out_left)
        supportRequestWindowFeature(Window.FEATURE_NO_TITLE)
        activityLoginBinding = DataBindingUtil.setContentView(this, R.layout.activity_login)

        initializeViewModels()
        initializeListeners()
    }

    private fun initializeListeners() {

        //        bindingLoginFragment.etKitchenId.doAfterTextChanged {
//
//            val isValidated = loginViewModel.validateUserName()
//            setErrorProperties(bindingLoginFragment.txtInputKitchenIdLayout,bindingLoginFragment.etKitchenId,!isValidated)
//        }
//
//        bindingLoginFragment.etPassword.doAfterTextChanged {
//            val isValidated = loginViewModel.validatePassword()
//            bindingLoginFragment.txtInputPasswordLayout?.let { it1 -> setErrorProperties(it1,bindingLoginFragment.etPassword,!isValidated) }
//        }

      //  activityLoginBinding.etWaiterId.setOnFocusChangeListener()
        activityLoginBinding.btnLogin.setOnClickListener { onClick(ClickType.Login) }
        activityLoginBinding.tvForgotPassword.setOnClickListener { onClick(ClickType.ForgotPassword) }
    }


    // Variable initialization
    private fun initializeViewModels() {
        activityLoginBinding.let { FragmentUtils.setBindingVariables(hashMapOf(BR.LoginVM to loginViewModel, BR.clickEvent to this), it) }
        activityLoginBinding.layoutLoader.apiErrorVM = apiErrorViewModel
        activityLoginBinding.layoutError.apiErrorVM = apiErrorViewModel

        activityLoginBinding.setLifecycleOwner(this)

    }

    private fun setErrorProperties(textInputLayout: TextInputLayout, textInputEditText: TextInputEditText, isError: Boolean) {
        setHintColor(textInputLayout,isError)
        textInputLayout.endIconDrawable = getEndIcon(isError)
        textInputLayout.isEndIconVisible = true

    }

    private fun setHintColor(view: TextInputLayout, isError: Boolean) {

        if (isError) view.markRequiredInRed()
        view.hintTextColor = getHintTextColor(isError)

        view.boxStrokeColor = getBoxStrokeColor(isError) ?: R.color.colorWaiterIdText
    }

    private fun getEndIcon(isError: Boolean) : Drawable? {
        return this.let { context ->
            if(isError) AppCompatResources.getDrawable(context,R.drawable.ic_error_red_24dp) else AppCompatResources.getDrawable(context,R.drawable.ic_done_tick_green_24dp)

        }
    }


    private fun getHintTextColor(isError: Boolean): ColorStateList? {
        return this.let { context ->
            if (isError) ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorCountdownTimer)) else ColorStateList.valueOf(ContextCompat.getColor(context, R.color.colorTabItemTextSelected))
        }
    }

    private fun getBoxStrokeColor(isError: Boolean): Int? {
        return this.let { context ->
            if (isError) ContextCompat.getColor(context, R.color.colorCountdownTimer) else ContextCompat.getColor(context, R.color.colorWaiterIdText)
        }

    }

    private fun loginUser() {
        val isValidated = loginViewModel.validateUserName() ?: false
        setErrorProperties( activityLoginBinding.txtInputPasswordLayout, activityLoginBinding.etWaiterId,!isValidated)

        val isPasswordValidated = loginViewModel.validatePassword() ?: false
        setErrorProperties(activityLoginBinding.txtInputPasswordLayout,activityLoginBinding.etPassword,!isPasswordValidated)


        FragmentUtils.hideKeyboard(activityLoginBinding.root,this)
        val deviceId = this.let { context -> SharedPreferenceManager(context, constant.LOGIN_SP).getStringPreference(constant.USER_DEVICEID) }

        val fcmToken = this.let { context -> SharedPreferenceManager(context, constant.LOGIN_SP).getStringPreference(constant.USER_FCMTOKENID) }



        if (deviceId != null && fcmToken != null) {
            loginViewModel.login(deviceId, fcmToken)?.observe(this, Observer {
                it?.let { response ->
                    when (response.status) {
                        APIStatus.LOADING -> FragmentUtils.showProgress(parentView = activityLoginBinding.viewLogin, toShow = true, apiErrorViewModel = apiErrorViewModel, progressMessage = getString(R.string.wait), isSwipeRefreshed = false)

                        APIStatus.ERROR ->  showError(true,errorCode = response.errorCode ?: 405, errorMessage = Utility.getMessageOnErrorCode(response.errorCode, this))

                        APIStatus.SUCCESS -> {
                            if (response.data == null) showError(false,404,"")
                            else {
                                activityLoginBinding.viewLogin.visibility = View.VISIBLE
                                apiErrorViewModel?.resetValues()
                                handleLoginResponse(response.data)

                            }
                        }
                    }
                }
            })
        }



    }

    private fun handleLoginResponse(response: Any) {

        val sharedPreferenceManager = SharedPreferenceManager(context = this, name = constant.LOGIN_SP)
        (response as? LoginResponse).let { loginResponse ->
            sharedPreferenceManager.storeStringPreference(API.AUTHORIZATION, loginResponse?.token
                    ?: "")
        }
        sharedPreferenceManager.storeBooleanPreference(constant.KEY_IS_LOGGED_IN, true)
        launchHomeActivity()

    }

    private fun launchHomeActivity() {
        Utility.ShowToast(this,getString(R.string.login_success), Toast.LENGTH_LONG)
        //  startActivity(Intent(activity,HomeActivityNew::class.java))
        //activity?.finish()
    }

    override fun onClick(clickType: ClickType) {
        if (clickType == ClickType.Login) loginUser()
    }

    fun TextInputLayout.markRequiredInRed() {

        hint = if (hint?.contains("*") == true) hint?.substring(0, (hint?.length
                ?: 0) - 1) else hint?.substring(0, hint?.length ?: 0)
        hint = buildSpannedString {

            append(hint)
            color(Color.RED) { append("*") } // Mind the space prefix.
        }

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
            hintTextColor = ColorStateList.valueOf(resources.getColor(R.color.colorCountdownTimer, null))
        }
    }

    fun showError(toShow: Boolean, errorCode: Int, errorMessage: String?) {

        if (toShow) apiErrorViewModel.let { FragmentUtils.showError( it, errorCode, errorMessage, true) } else apiErrorViewModel?.resetValues()
        activityLoginBinding.viewLogin.visibility = if (toShow) View.GONE else View.VISIBLE
        if (toShow) {
            Handler().postDelayed({ showError(false, errorCode, "") }, 3000)
        }


    }

}